### ISSUE TITLE
Should parents only refer to published modules?

### ISSUE THREAD

**Issue by jameslibscie** [p2pcommons/specs#27]
This issue follows from a discussion on [@hypergraph-xyz/desktop/#46](https://github.com/hypergraph-xyz/desktop/pull/46).

In theory, when publishing as-you-go, all prior work is already published and `parents` should only be able to refer to published work. However, in the case that a content creator does want to work on multiple consecutive modules as drafts (or is creating content based on a colleague's draft), this is then prevented which could cause users to abandon the `parents` concept even where it is useful.

Suggestion: How about we require that *published* modules' `parents` must be published, but until then, it is allowed to refer to drafts?

@chartgerink @juliangruber @dpaez

**Comment by juliangruber** [comment]
> Suggestion: How about we require that _published_ modules' `parents` must be published, but until then, it is allowed to refer to drafts?

Me like!

**Comment by chartgerink** [comment]
Circling back to this issue, here I write out some scenario's within the p2pcommons space, specifically for our Hypergraph implementation:

# Scenario 1

Hypergraph user creates a follow up to a content module that has already been added to a profile. 

This sets the parent for the newly drafted content module automatically and does not require us to expose any versioning. 

The chain of content versions is set except for the new addition, which gets set upon adding to profile. 

*This means one cannot start a draft on a module that has not been added to a profile. This enforces as-you-go reporting.*

# Scenario 2

Hypergraph user adds a draft content module with a draft content module as parent.

When the user wants to add the content module to their profile, this will require the user to cascade into adding the parent module of the draft before being able to add to profile.

In the worst case scenario, we can run the risk of parents of the draft being drafts themselves to the Nth degree. In other words, a draft parent is a draft whose parent is a draft etc. This would require all of those to be added to profile one by one with a potential risk of breaking the chain of versioning (more room for mistakes after all).

```
theory-draft
└── predictions-draft
    └── design-draft
        └── data-draft
            └── code-draft
                └── results-draft
```

In this situation, the `results-draft` could only be added to profile if `code-draft` is already added, etc.

We could hypothetically include a "Add chain to profile" but that would also open up the possibility of adding a chain after-the-fact, sort of making the whole point of as-you-go moot. That is my main concern for Scenario 2, that it opens the door to after-the-fact reporting whereas that's what we're exactly trying to prevent.

# Interim conclusion

One can argue Scenario 2 is just the responsibility of the Hypergraph user. Scenario 2 does allow for more flexibility in how work is done, but is potentially so lenient it undermines the work we're trying to do by allowing after-the-fact reporting.

Scenario 1 is the conservative approach and any relaxation of that spec would *not* be a breaking change, but going from Scenario 2 to 1 would be. Additionally, Scenario 1 is in line with as-you-go reporting. Part of the stepwise research communication process is also that you don't have to keep thinking about all the steps, but just the latest, which would fit better with a Scenario 1 as well, I'd argue.

It would greatly simplify checks and user flows if we simply opt for Scenario 1 and require users to only connect a next link in the chain to content modules that have already been added to a profile. We could then opt for revisiting this in the spec if usage deems it necessary, with the connotation that this runs the risk of reintroducing after-the-fact practices. 

**tl;dr Yes, parents must be added to profile at the time of creating a new content module**

**Comment by juliangruber** [comment]
> We could hypothetically include a "Add chain to profile" but that would also open up the possibility of adding a chain after-the-fact, sort of making the whole point of as-you-go moot. That is my main concern for Scenario 2, that it opens the door to after-the-fact reporting whereas that's what we're exactly trying to prevent.

This is a very important point. We need to guide our users to behave as beneficial to them and the system as possible.

**Comment by jameslibscie** [comment]
I agree with what's been said so far. Let's do it :)
